### ISSUE TITLE
MotionStreak causes crashes sometimes (more details in the post)

### ISSUE THREAD

**Issue by iPruch** [cocos2d/cocos2d-x#14149]
Hello,

I am currently developing a game with Cocos2d-x 3.7 (updated the project to 3.8.1 to check if the problem would magically go away, but it didn't. Same with cocos2d-x 3.9). The game is a kind of endless runner, and I am using the `cocos2d::MotionStreak` class to add a nice effect to the coins collected by the player. Coins travel to the coins counter when the player collides with them, and each coin has a MotionStreak effect.

You can see a video of the game here: https://www.youtube.com/watch?v=Kaxf1Imu600

And a screenshot:
![a62f47b4-7d5b-11e5-9b6d-c598ba90c374](https://cloud.githubusercontent.com/assets/2997598/10784073/437916bc-7d5c-11e5-986e-325c6e8bad4d.png)

I've been developing and testing the game for a couple of months and everything seemed fine. But someday I decided to dramatically increase the speed of the game, as a stress test or just for fun. But after some seconds of gameplay, a crash occurred: 

**Crash 1:**

`CCASSERT(_glprogram, "invalid glprogram");` <- EXC_BAD_ACCESS

![caf4a85340fe894e](https://cloud.githubusercontent.com/assets/2997598/10507890/18acbcbe-7324-11e5-8cd6-427eab3bca8c.png)

I do not know how I ended suspecting of the MotionStreak effect, but I tried to comment the lines where I created and added the MotionStreak effect to each coin. And after a lot of games, I did not manage to reproduce the crash. To confirm that it was something about the MotionStreak, I enabled the MotionStreak effect again and it brought the crashes back.

I found it was a lot more frequent on older devices, such as the iPhone 4 or iPad mini 1, than on newer devices, like the iPod Touch 6. ~~I haven't tested it on Android as debugging is more difficult as I only want to use Xcode~~. **Edit:** I tested it on Android devices with same results: powerful devices such as LG G3 crash a lot less frequently than mid/low tier devices.

I also got other less frequent crashes:

**Crash 2:**

`CCASSERT(command->getType() != RenderCommand::Type::UNKNOWN_COMMAND, "Invalid Command Type");` <- Signal SIGABRT (the assertion did not succeed)

![310eaea3426dfe8c](https://cloud.githubusercontent.com/assets/2997598/10507908/2add1a0a-7324-11e5-960b-feda69c4fb6a.png)

---

**Crash 3:**

`cocos2d::Renderer::fillQuads(cocos2d::QuadCommand *cmd)` <- EXC_BAD_ACCESS inside this function 

![cee1a9964f2831b2](https://cloud.githubusercontent.com/assets/2997598/10507915/3a0f7f22-7324-11e5-8ed5-3b63cfd6be01.png)

---

**Crash 4:**

`CCASSERT(cmd->getQuadCount()>= 0 && cmd->getQuadCount() * 4 < VBO_SIZE, "VBO for vertex is not big enough, please break the data down or use customized render command");` <- Signal SIGABRT  (the assertion did not succeed)

![a4274367402a157a](https://cloud.githubusercontent.com/assets/2997598/10507917/3f5c00cc-7324-11e5-9680-630a1a8ca4e6.png)

---

You could say: "Hey, but your game ran perfectly before increasing the gameplay speed. Why don't you just set it as before?"

**The problem** is that the game's velocity slowly increases in order to become more difficult. Besides, I was planning to create a power up that boosts the player a lot. Not to mention the quantity of different compatible devices which could behave differently and crash at lower speeds.

I don't know, but it could be related to the way MotionStreak nodes load textures (for some reason I can't correctly use the sprite sheet here, as it shows the whole texture instead of only the portion corresponding to the trail texture), as I have to specify the texture apart from the sprite sheet:

`auto motionStreak = MotionStreak::create(fade, minSeg, stroke, Color3B(220, 255, 124), "trail.png");`

Any light on these dark MotionStreak crashes?

**Comment by iPruch** [comment]
Anyone, please? @minggo @ricardoquesada 

I know that post is long and you guys have little time, but I'm really stuck with this. I'd be really thankful to you.

**Comment by jgcoded** [comment]
I don't know the answer to your problem, but I am using MotionStreak in my game. I'm using cocos2d-x v3.9 and so far I haven't experienced any crashes. Maybe v3.9 fixed it? Try it and let me know

**Comment by ricardoquesada** [comment]
@iPruch Thanks for the detailed bug report, but I will need a test case... a project that reproduces the bug. Otherwise it will be very difficult to fix it.

**Comment by iPruch** [comment]
@jgcoded I updated the project to v3.9 some weeks ago, but sadly crashes are still present.

@ricardoquesada I'd create a test case or sample project if I knew what on the world is happening there...  But I'll try and come back here if I manage to reproduce it.

Thank you both for your suggestions!

**Comment by iPruch** [comment]
Well... I don't know what has changed between the last time I tested it and now, but I'm unable to reproduce these crashes again. Unfortunately, now I'm getting crashes related to the particle system, which I didn't experienced before. I replaced the ParticleSystem classes of the project with Cocos2d-x 3.8.1 versions and now it all seem to work.
I'd really like to know what was causing those crashes, but I'm running out of time with this project. I hope they don't appear again, and if so I'll come back here.

Thank you for your time!
